### PR TITLE
Component-ize snowball generation

### PR DESCRIPTION
**Pull request by Azrun**
## About the PR 
https://streamable.com/j964g3

- Component-ize snowball generation to allow it to be placed on an arbitrary turf.

Snow Ball Changes:
- Snowballs now use an actionbar which lasts 3-5 seconds, and the turf can not be used again for 6 seconds.
- Snowballs will take on any color of the originating turf.

COMSIG Changes:
- Adds COMSIG_ATTACKHAND for all your attack_hand capturing needs.

Actionbar Changes:
- Updated the base callback Action bar to have a new parameter to specify a new datum to be the target of the callback call. (So components can use the callback actionbar without having to duplicate logic)

## Why's this needed? 
I want this for _Artemis_.  Provides groundwork for use of `attack_hand` support by components.
_Oh... and winterify_

### CODE REVIEW

File: code/turf/floors.dm [hunk 1]
```diff
@@ -1028,18 +1028,8 @@ DEFINE_FLOORS(marble/border_wb,
...
-			return
-		else
-			user.visible_message("<b>[user]</b> gathers up some snow and rolls it into a snowball!",\
-			"You gather up some snow and roll it into a snowball!")
-			var/obj/item/reagent_containers/food/snacks/snowball/S = new /obj/item/reagent_containers/food/snacks/snowball(user.loc)
-			user.put_in_hand_or_drop(S)
-			src.last_gather_time = world.time
-			return
+	New()
+		AddComponent(/datum/component/snowballs)
```
Comment (BatElite) [inline code/turf/floors.dm:1032]: Linter's complaining you're not calling the parent proc here

File: code/datums/limb.dm [hunk 2]
```diff
@@ -37,7 +37,8 @@
 	proc/attack_hand(atom/target, var/mob/user, var/reach, params, location, control)
 		if(!target) // fix runtime Cannot execute null.attack hand().
 			return
-
+		if(SEND_SIGNAL(target, COMSIG_ATTACKHAND, user))
+			return
```
Comment (Tarmunora) [inline code/datums/limb.dm:41]: There is no way in hell that this catches all `atom/proc/attack_hand` calls 😭 
Will probably need a similar garbage a la `Attackby()` to make sure the signal is always sent whenever `atom/proc/attack_hand()` is called...

**Comment by Azrun** [comment]
Wanted to surface some concerns:
image

image

**Comment by Azrun** [comment]
Follow on PR Idea: Have the reagent coloration actually be a chem-based where the snowballs would have `ice`, take in N units of the fluid group on creation and if the temperature causes the ice to melt you lose the snowball.  The hit proc could then apply the on_touch() reagent reaction to target.

File: code/datums/components/snowballs.dm [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+		list(user), 'icons/misc/xmas.dmi', "snowball", null, null, src), user)
+		return TRUE
+
+/datum/component/snowballs/proc/form_snowball(mob/user)
+	user.visible_message("<b>[user]</b> gathers up some snow and rolls it into a snowball!", "You gather up some snow and roll it into a snowball!")
+	var/obj/item/reagent_containers/food/snacks/snowball/S = new /obj/item/reagent_containers/food/snacks/snowball(user.loc)
+	var/datum/reagents/R = source_turf.active_liquid?.group?.reagents
+	if(R)
+		S.color = R.get_average_color().to_rgb()
+	else
```
Comment (Tarmunora) [inline code/datums/components/snowballs.dm:27]: maybe let's not let people piss on the floor and make yellow snowballs...

Comment (Azrun) [reply code/datums/components/snowballs.dm:27]: But... but.... okay.

Comment (Azrun) [reply code/datums/components/snowballs.dm:27]: People.  People are why we can't have gross things.

Comment (Azrun) [reply code/datums/components/snowballs.dm:27]: ~~Maybe let them do it, but if they do it more than once a rabid husky jumps from the snowpile and rips them limb from limb.~~

Removing.

**Pull request merged**